typing/printtyp.ml

```diff
@@ -409,22 +413,35 @@ let add_named_var ty =
       named_vars := name :: !named_vars
   | _ -> ()
 
-let rec new_name () =
+let rec new_name weak () =
   let name =
-    if !name_counter < 26
-    then String.make 1 (Char.chr(97 + !name_counter))
-    else String.make 1 (Char.chr(97 + !name_counter mod 26)) ^
-           string_of_int(!name_counter / 26) in
-  incr name_counter;
-  if List.mem name !named_vars
+    if weak = None then
+      if !name_counter < 26
+      then String.make 1 (Char.chr(97 + !name_counter))
+      else String.make 1 (Char.chr(97 + !name_counter mod 26)) ^
+           string_of_int(!name_counter / 26)
+    else
+      "weak" ^ string_of_int !weak_counter
+  in
+  if weak = None then incr name_counter else incr weak_counter;
+  if List.mem name !named_vars || StringSet.mem name !named_weak_vars
   || List.exists (fun (_, name') -> name = name') !names
-  then new_name ()
-  else name
+  then new_name weak ()
+  else (
+    begin match weak with
+    | Some t ->
+        named_weak_vars := StringSet.add name !named_weak_vars;
+        weak_var_map := TypeMap.add t name !weak_var_map
+    | None -> ()
+    end;
+    name
+  )
```

Comment:
Splitting the functions in two seemed indeed clearer (I may have changed few names around).
Concerning `new_name`, new_weak_name and new_name always tend to appear in pair, so it did not seem very useful to precise  that `new_name`  is `new_not_weak_name`.